hybridse/src/plan/planner.cc

```diff
@@ -1118,6 +1134,136 @@ base::Status Planner::TransformTableDef(const std::string &table_name, const Nod
     return base::Status::OK();
 }
 
+std::map<const node::WindowDefNode *, node::ProjectListNode *> Planner::FilterWindowFrameBoundNonRelativeProjects(
+    const std::map<const node::WindowDefNode *, node::ProjectListNode *> &in, int *win_id) {
+    std::map<const node::WindowDefNode *, node::ProjectListNode *> splited_map;
+
+    for (auto& kv : in) {
+        if (kv.first == nullptr) {
+            splited_map.emplace(kv.first, kv.second);
+            continue;
+        }
+
+        if (kv.first->GetFrame() == nullptr || !kv.first->GetFrame()->IsPureHistoryFrame()) {
+            // only split project list if the window is pure history
+            splited_map.emplace(kv.first, kv.second);
+            continue;
+        }
+
+        node::ProjectListNode* window_bound_relative_projects = nullptr;
+        node::ProjectListNode* window_bound_non_relative_projects = nullptr;
+        SplitProjectList(
+            kv.second,
+            [](const node::ProjectNode *n) {
+                /* lag functions actually do not rely on window frame bound, the predicate will filter those out */
+                if (n == nullptr) {
+                    return true;
+                }
+
+                if (n->GetExpression()->GetExprType() != node::kExprCall) {
+                    return true;
+                }
+
+                auto *call_expr = dynamic_cast<node::CallExprNode *>(n->GetExpression());
+                if (call_expr == nullptr) {
+                    return true;
+                }
+                const auto &fn_name = call_expr->GetFnDef()->GetName();
+                return !absl::EqualsIgnoreCase("lag", fn_name) && !absl::EqualsIgnoreCase("at", fn_name) &&
+                       !absl::EqualsIgnoreCase("lead", fn_name);
+            },
+            &window_bound_relative_projects, &window_bound_non_relative_projects);
+
+        if (window_bound_non_relative_projects->GetProjects().empty()) {
+            // there is no lag style functions in current project list node, keep it as-is
+            splited_map.emplace(kv.first, kv.second);
+        } else {
+            // found lag/at style functions
+
+            bool preserve_old_window = !window_bound_relative_projects->GetProjects().empty();
+
+            if (preserve_old_window) {
+                splited_map.emplace(kv.first, window_bound_relative_projects);
+            }
+
+            auto* window_bound_non_relative_window = CreateWindowUnboundPrecedingAndCurrent(kv.first);
+            // prohibit window merge on the window
+            window_bound_non_relative_window->SetPermitWindowMerge(false);
```

Comment:
maybe we can verify whether this is necessary. Further merging should work properly.

Reply:
yeah, I removed this